general/mem_manager.hpp

```diff
@@ -13,33 +13,462 @@
 #define MFEM_MEM_MANAGER_HPP
 
 #include "globals.hpp"
+#include "error.hpp"
+#include <cstring> // std::memcpy
+#include <type_traits> // std::is_const
 
 namespace mfem
 {
 
 // Implementation of MFEM's lightweight device/host memory manager designed to
 // work seamlessly with the OCCA, RAJA, and other kernels supported by MFEM.
 
+/// Memory types supported by MFEM.
+enum class MemoryType
+{
+   HOST,      ///< Host memory; using new[] and delete[]
+   HOST_32,   ///< Host memory alligned at 32 bytes
+   HOST_64,   ///< Host memory alligned at 64 bytes
+   CUDA,      ///< cudaMalloc, cudaFree
+   CUDA_UVM   ///< cudaMallocManaged, cudaFree
+};
+
+/// Memory classes identify subsets of memory types.
+/** This type is used by kernels that can work with multiple MemoryType%s to
+    request access to any input/output Memory objects. */
+enum class MemoryClass
+{
+   HOST,    ///< Memory types: { HOST, HOST_32, HOST_64, CUDA_UVM }
+   HOST_32, ///< Memory types: { HOST_32, HOST_64 }
+   HOST_64, ///< Memory types: { HOST_64 }
+   CUDA,    ///< Memory types: { CUDA, CUDA_UVM }
+   CUDA_UVM ///< Memory types: { CUDA_UVM }
+};
+
+/// Return true is the given memory type is in MemoryClass::HOST.
+inline bool IsHostMemory(MemoryType mt) { return mt <= MemoryType::HOST_64; }
+
+/// Return a suitable MemoryType for a given MemoryClass.
+MemoryType GetSuitableMemoryType(MemoryClass mc);
+
+/// Return a suitable MemoryClass from a pair of MemoryClass%es.
+/** Note: this operation is commutative, i.e. a*b = b*a, associative, i.e.
+    (a*b)*c = a*(b*c), and has an identity element: MemoryClass::HOST.
+
+    Currently, the operation is defined as a*b := max(a,b) where the max
+    operation is based on the enumeration ordering:
+
+        HOST < HOST_32 < HOST_64 < CUDA < CUDA_UVM. */
+MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
```

Comment:
This operator was very confusing when looking at the source code. Why is this not a regular function that is more descriptive?

```cpp
GetSuitableMemoryClass(MemoryClass a, MemoryClass b);
```

Reply:
With the overloaded `operator*()` it is much shorter to write compositions: `A*B*C*D` 😃 . Initially, I had a function like what you suggest but then I switched it. I tend to agree that the regular function is more descriptive. Should I switch to that?

Reply:
We could always overload it to have more inputs hehe. I would prefer to switch it so it's more understandable at a glance.

Reply:
Variadic templates would work too, but I would understand that we don't want to use them here.